crates/fuzzing/src/oom.rs

```diff
@@ -0,0 +1,211 @@
+                }
+            }
+        }
+
+        set_oom_state(new_state);
+        ptr
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: std::alloc::Layout) {
+        unsafe {
+            std::alloc::System.dealloc(ptr, layout);
+        }
+    }
+}
+
+/// A test helper that checks that some code handles OOM correctly.
+///
+/// `OomTest` will only work correctly when `OomTestAllocator` is configured as
+/// the global allocator.
+///
+/// `OomTest` does not support reentrancy, so you cannot run an `OomTest` within
+/// an `OomTest`.
+///
+/// # Example
+///
+/// ```no_run
+/// use std::time::Duration;
+/// use wasmtime::Result;
+/// use wasmtime_fuzzing::oom::{OomTest, OomTestAllocator};
+///
+/// #[global_allocator]
+/// static GLOBAL_ALOCATOR: OomTestAllocator = OomTestAllocator::new();
+///
+/// #[test]
+/// fn my_oom_test() -> Result<()> {
+///     OomTest::new()
+///         .max_iters(1_000_000)
+///         .max_duration(Duration::from_secs(5))
+///         .test(|| {
+///             todo!("insert code here that should handle OOM here...")
+///         })
+/// }
+/// ```
+pub struct OomTest {
+    max_iters: Option<u32>,
+    max_duration: Option<time::Duration>,
+}
+
+impl OomTest {
+    /// Create a new OOM test.
+    ///
+    /// By default there is no iteration or time limit, tests will be executed
+    /// until the pass (or fail).
+    pub fn new() -> Self {
+        let _ = env_logger::try_init();
+        OomTest {
+            max_iters: None,
+            max_duration: None,
+        }
+    }
+
+    /// Configure the maximum number of times to run an OOM test.
+    pub fn max_iters(&mut self, max_iters: u32) -> &mut Self {
+        self.max_iters = Some(max_iters);
+        self
+    }
+
+    /// Configure the maximum duration of time to run an OOM text.
+    pub fn max_duration(&mut self, max_duration: time::Duration) -> &mut Self {
+        self.max_duration = Some(max_duration);
+        self
+    }
+
+    /// Repeatedly run the given test function, injecting OOMs at different
+    /// times and checking that it correctly handles them.
+    ///
+    /// The test function should return an `Err(_)` if and only if it encounters
+    /// an OOM.
+    ///
+    /// Returns early once the test function returns `Ok(())` before an OOM has
+    /// been injected.
+    pub fn test(&self, test_func: impl Fn() -> Result<()>) -> Result<()> {
+        let start = time::Instant::now();
+
+        for i in 0.. {
+            if self.max_iters.is_some_and(|n| i >= n)
+                || self.max_duration.is_some_and(|d| start.elapsed() >= d)
+            {
+                break;
+            }
+
+            log::trace!("=== Injecting OOM after {i} allocations ===");
+            let old_state = set_oom_state(OomState::OomOnAlloc(i));
+            assert_eq!(old_state, OomState::OutsideOomTest);
+
+            let result = test_func();
+            let old_state = set_oom_state(OomState::OutsideOomTest);
```

Comment:
This'll likely want to be in a `Drop` to avoid heavily crashing the process by accident in the case that `test_func` panics